server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.store;
+
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.FSDirectory;
+import org.apache.lucene.store.FilterDirectory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+import org.apache.lucene.store.Lock;
+import org.opensearch.index.store.remote.file.OnDemandCompositeBlockIndexInput;
+import org.opensearch.index.store.remote.filecache.CachedIndexInput;
+import org.opensearch.index.store.remote.filecache.FileCache;
+import org.opensearch.index.store.remote.utils.filetracker.FileState;
+import org.opensearch.index.store.remote.utils.filetracker.FileType;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Set;
+
+public class CompositeDirectory extends FilterDirectory {
+
+    private final FSDirectory localDirectory;
+    private final RemoteStoreFileTrackerAdapter remoteStoreFileTrackerAdapter;
+    private final FileCache fileCache;
```

Comment:
Do we have eviction algorithm for this cache?

Reply:
The short answer is yes, the longer answer is that this is kind of a weird "cache". It points to files on disk. If the files on disk are actively being used by on-going search operations, then they are not eligible for eviction. As soon as they become unreferenced, they are kept around as long as there is space and evicted with an LRU algorithm. You can dig into this more here: https://github.com/opensearch-project/OpenSearch/blob/a1e42b1131d5d807dc66d13dd960ebe707da3f09/server/src/main/java/org/opensearch/index/store/remote/utils/cache/LRUCache.java#L28-L41